### PR TITLE
Add support IAM policy for the Environment of Apigee X

### PR DESCRIPTION
**Pull request by xuchenma**
This PR fixes https://github.com/hashicorp/terraform-provider-google/issues/9767 by adding IAM support to ApigeeEnvironment.
It also changed the ApigeeEnvironment resource from requiring "org_id = organizations/<org_name>" to "org_name = <org_name>".
It also refactored some tests for ApigeeOrganization and ApigeeEnvironment. Previously there were separated sets of tests, some written solely for docs (skipped testing) and some solely for testing (skipped docs). This PR used the same test for docs and testing wherever possible. For cases where it has to be separated tests, it minimized the difference between them.

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [X] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
apigee: Added IAM support for `google_apigee_environment`.
```

```release-note:breaking-change
apigee: Replaced `org_id` with `org_name` for `google_apigee_environment`.
```

### CODE REVIEW

**Comment by xuchenma** [comment]
I ran the following acceptance tests in my local:
TestAccApigeeOrganization_apigeeOrganizationCloudBasicExample
TestAccApigeeOrganization_apigeeOrganizationCloudFullTestExample (in beta)
TestAccApigeeEnvironment_apigeeEnvironmentBasicExample
TestAccApigeeEnvironmentIamPolicyGenerated
TestAccApigeeEnvironmentIamBindingGenerated
TestAccApigeeEnvironmentIamMemberGenerated

5 out of the 6 passed, except for "TestAccApigeeEnvironmentIamPolicyGenerated".
The cause seemed to be that, the payload for the POST call to "/v1/organizations/<apigee-org>/environments/<apigee-env>:setIamPolicy" did not include the `etag` field. It gave 400 with error message: "Validation error: Etag must be set during SetPolicy and PurgePolicy calls. You can obtain current etag value for this resource by making a GetPolicy call first."
However during "TestAccApigeeEnvironmentIamBindingGenerated" and "TestAccApigeeEnvironmentIamMemberGenerated", the POST calls were made with `etag` in the payload. I'm not sure what I can do to fix this.

**Comment by xuchenma** [comment]
Thank you for the reply, Stephen!
I made a few more commits. Right now the api.yaml only has minimal code changes to support IAM.

I ran the following tests in my local:
TestAccApigeeOrganization_apigeeOrganizationCloudBasicExample => PASSED
TestAccApigeeOrganization_apigeeOrganizationCloudFullTestExample (in beta) => PASSED
TestAccApigeeEnvironment_apigeeEnvironmentBasicExample => PASSED
TestAccApigeeEnvironmentIamPolicyGenerated => FAILED, policy cannot be POST'ed for the same "etag" reason as in my previous comment.
TestAccApigeeEnvironmentIamBindingGenerated => FAILED, policy can be POST'ed, but GET afterwards fetched the wrong URL endpoint. It should fetch v1/organizations/<org_name>/environments/<env_name>; however it fetches v1/<org_name>/environments/<env_name> (missed the "organizations/" part).
TestAccApigeeEnvironmentIamMemberGenerated => FAILED for the same reason, missed the "organizations/" part.

Because I only made minimal changes to api.yaml, I'm not sure what I'm missing here. Please advise. Thanks!

**Comment by melinath** [comment]
It looks like the reason that etag is present for the member/binding resources and not for the policy resources is because member/binding need to do a get request to merge the existing IAM policy with their changes - which gets them an etag as a side effect. The policy resource overrides whatever's there so it doesn't do a get request first (and so doesn't get an  etag). I think we expect this kind of request to go through; this API must be slightly different from standard.

**Comment by melinath** [comment]
The issues with the member/binding resources have to do with the import URL. I'm not entirely sure what the fix will be; I need to look into it some more.

File: mmv1/products/apigee/api.yaml [hunk 1]
```diff
@@ -231,6 +231,13 @@ objects:
         message: 'message'
     description: |
       An `Environment` in Apigee.
+    iam_policy: !ruby/object:Api::Resource::IamPolicy
+      exclude: false
+      method_name_separator: ':'
+      parent_resource_attribute: 'env_id'
+      import_format: ["{{org_id}}/environments/{{name}}", "{{name}}"]
```
Comment (melinath) [inline mmv1/products/apigee/api.yaml:238]: This `%` will indicate that the org_id can contain slashes, which should resolve the problem. (Currently the import_format regex excludes the "organization/" at the start of the org id.)

```suggestion
      import_format: ["{{%org_id}}/environments/{{name}}", "{{name}}"]
```

**Review by melinath — CHANGES_REQUESTED** [review]
This should resolve the issues with member/binding.

**Comment by xuchenma** [comment]
Thanks! I just committed the suggested change and rebased.

**Comment by melinath** [comment]
run of just new tests: https://ci-oss.hashicorp.engineering/buildConfiguration/GoogleCloudBeta_ProviderGoogleCloudBetaMmUpstream/216659

File: mmv1/products/apigee/terraform.yaml [hunk 2]
```diff
@@ -103,6 +103,7 @@ overrides: !ruby/object:Overrides::ResourceOverrides
       # the resources needed for the acceptance test.
       name: "apigee_environment_basic_test"
       primary_resource_id: "apigee_environment"
+      primary_resource_name: "getTestProjectFromEnv(), fmt.Sprintf(\"tf-test-apigee-env%s\", context[\"random_suffix\"])"
```
Comment (melinath) [inline mmv1/products/apigee/terraform.yaml:106]: It looks like this is causing issues; it is causing things like this to be built:

```
			{
				ResourceName:      "google_apigee_environment_iam_binding.foo",
				ImportStateId:     fmt.Sprintf("%s/environments/%s roles/viewer", getTestProjectFromEnv(), fmt.Sprintf("tf-test-apigee-env%s", context["random_suffix"])),
				ImportState:       true,
				ImportStateVerify: true,
			},
```

This results in the import id being set to something like `my-project/environments/tf-test-apigee-envasdfklj` but it should be `organizations/tf-testasdfklj/environments/tf-testasdfklj`. I think you could resolve the issue with something like this:

```suggestion
      primary_resource_name: "fmt.Sprintf(\"organizations/tf-test%s\", context[\"random_suffix\"]), fmt.Sprintf(\"tf-test%s\", context[\"random_suffix\"])"
```

Comment (melinath) [reply mmv1/products/apigee/terraform.yaml:106]: Let me know if that helps the tests pass for you locally!

**Review by melinath — CHANGES_REQUESTED** [review]

File: mmv1/products/apigee/terraform.yaml (hunk 2 above)
Comment (xuchenma) [reply mmv1/products/apigee/terraform.yaml]: Thanks, done! I'm still working with my team to setup the permission for me to run tests locally. We can try and see if the tests pass in your system.

**Comment by melinath** [comment]
Run of the IAM tests: https://ci-oss.hashicorp.engineering/buildConfiguration/GoogleCloudBeta_ProviderGoogleCloudBetaMmUpstream/217239

**Comment by melinath** [comment]
It looks like the binding and member tests are now passing, so all that's left is the policy test.

**Comment by melinath** [comment]
Run of just the IAM tests: https://ci-oss.hashicorp.engineering/buildConfiguration/GoogleCloudBeta_ProviderGoogleCloudBetaMmUpstream/221390

**Review by melinath — APPROVED** [review]
LGTM

**Pull request merged**